Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -413,6 +422,133 @@ DocumentKey Serializer::DecodeKey(absl::string_view name) const {
+                                const ObjectValue& object_value) const {
+  // Encode Document.name
+  writer->WriteTag({PB_WT_STRING, google_firestore_v1beta1_Document_name_tag});
+  writer->WriteString(EncodeKey(key));
+
+  // Encode Document.fields (unless it's empty)
+  if (!object_value.internal_value.empty()) {
+    writer->WriteTag(
+        {PB_WT_STRING, google_firestore_v1beta1_Document_fields_tag});
+    EncodeObject(writer, object_value);
+  }
+
+  // Skip Document.create_time and Document.update_time, since they're
+  // output-only fields.
+}
+
+util::StatusOr<std::unique_ptr<model::MaybeDocument>>
+Serializer::DecodeMaybeDocument(const uint8_t* bytes, size_t length) const {
+  Reader reader = Reader::Wrap(bytes, length);
+  std::unique_ptr<MaybeDocument> maybeDoc =
+      DecodeBatchGetDocumentsResponse(&reader);
+
+  if (reader.status().ok()) {
+    return maybeDoc;
+  } else {
+    return reader.status();
+  }
+}
+
+std::unique_ptr<MaybeDocument> Serializer::DecodeBatchGetDocumentsResponse(
+    Reader* reader) const {
+  Tag tag = reader->ReadTag();
+  if (!reader->status().ok()) return nullptr;
+
+  // Ensure the tag matches the wire type
+  switch (tag.field_number) {
+    case google_firestore_v1beta1_BatchGetDocumentsResponse_found_tag:
+    case google_firestore_v1beta1_BatchGetDocumentsResponse_missing_tag:
+      if (tag.wire_type != PB_WT_STRING) {
+        reader->set_status(
+            Status(FirestoreErrorCode::DataLoss,
+                   "Input proto bytes cannot be parsed (mismatch between "
+                   "the wiretype and the field number (tag))"));
+      }
+      break;
+
+    default:
+      reader->set_status(Status(
+          FirestoreErrorCode::DataLoss,
+          "Input proto bytes cannot be parsed (invalid field number (tag))"));
+  }
+
+  if (!reader->status().ok()) return nullptr;
+
+  switch (tag.field_number) {
+    case google_firestore_v1beta1_BatchGetDocumentsResponse_found_tag:
+      return reader->ReadNestedMessage<std::unique_ptr<MaybeDocument>>(
+          nullptr, [this](Reader* reader) -> std::unique_ptr<MaybeDocument> {
+            return DecodeDocument(reader);
+          });
+    case google_firestore_v1beta1_BatchGetDocumentsResponse_missing_tag:
+      // TODO(rsgowman): Right now, we only support Document (and don't support
+      // NoDocument). That should change in the next PR or so.
+      abort();
+    default:
+      // This indicates an internal error as we've already ensured that this is
+      // a valid field_number.
+      FIREBASE_ASSERT_MESSAGE(
+          false,
+          "Somehow got an unexpected field number (tag) after verifying that "
+          "the field number was expected.");
+  }
+}
+
+std::unique_ptr<Document> Serializer::DecodeDocument(Reader* reader) const {
+  std::string name;
+  FieldValue fields = FieldValue::ObjectValueFromMap({});
+  SnapshotVersion version = SnapshotVersion::None();
+
+  while (reader->bytes_left()) {
```

Comment:
If this is called with a reader whose status is not OK but which has no bytes left this will create an empty document rather than the result of `if (!reader->status().ok()) return nullptr;`